plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -19,16 +23,17 @@ import Control.Lens (ix, (^?))
 import Control.Monad.Except
 import Data.Array
 import Data.Kind qualified as GHC (Type)
-import PlutusCore.Builtin.KnownType
+import GHC.Exts (inline)
 
 -- | Same as 'TypeScheme' except this one doesn't contain any evaluation-irrelevant types stuff.
 data RuntimeScheme val (args :: [GHC.Type]) res where
     RuntimeSchemeResult
         :: KnownTypeIn (UniOf val) val res
         => RuntimeScheme val '[] res
     RuntimeSchemeArrow
-        :: KnownTypeIn (UniOf val) val arg
-        => RuntimeScheme val args res
+           -- 'readKnown' inlined for optimization purposes.
```

Comment:
So I guess the idea is that this is better than passing the class dictionary. What would happen if you split `KnownTypeIn` in two? Then you'd have single-method classes, which get represented directly as the underlying function. Perhaps that would also inline and specialize well.

Reply:
Then you could do the same trick for `RuntimeSchemeResult`, which I guess is also packing a dictionary instead of just taking the `makeKnown` function.

Reply:
I see, you explain this in the note below. Maybe we should investigate what goes wrong with the split more thoroughly.

Reply:
> Then you'd have single-method classes, which get represented directly as the underlying function. Perhaps that would also inline and specialize well.

According to my experiments, it's not necessarily the case, but I'll make sure to check once we perform the split (I really hope we'll do that in the end).